> > The possibility to return -1 from function `nlayer(int det, int part, int lay)` (which will never happen in practice, given its fixed input parameters) is generating build warnings in the IBs.
> 
> it may help to quote the warning. Apparently the issue is in the calling code not the actual implementation here.

Point taken: PR updated without modifying the implementation of the `nlayer` method

About the warning, it is the following:

```
>> Compiling  /scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/src/TrackerMap.cc
/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DEIGEN_DONT_PARALLELIZE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_13_3_X_2023-07-27-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_13_3_X_2023-07-27-2300' -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/pcre/8.43-37eb2e8b73bab83d6645ecfd5d73dcaa/include -isystem/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/boost/1.80.0-97c54cea73711453454f84250056342a/include -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/clhep/2.4.6.0-8b07d134e1b552a3656874209f04981b/include -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/cuda/12.0.1-c5280cf8b2a74b3d1a049eae475340f3/include -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/gsl/2.6-f316a321a173f181b66df52be79d894b/include -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/lcg/root/6.26.11-acd27c1cd56f6766f848290f955a0c7a/include -isystem/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/tbb/v2021.9.0-ed3b643ff5cfdeaf4a4df9ab15162172/include -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/zlib/1.2.11-51072030b7f93c3ac6c4235f21e413cb/include -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/eigen/82dd3710dac619448f50331c1d6a35da673f764a-b4fe5ef46e60c6c470cd12fab50ff23b/include/eigen3 -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/fmt/8.0.1-54e94b39f5cf29341bb9c4765764e1ca/include -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/OpenBLAS/0.3.15-15504a5c800a9ecbbc2befbb991bead5/include -I/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/el8_ppc64le_gcc12/external/tinyxml2/6.2.0-d17873b4d6a42a43226cf689f82ec1ef/include -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -fsigned-char -fsigned-bitfields -mcpu=power8 -mtune=power8 --param=l1-cache-size=64 --param=l1-cache-line-size=128 --param=l2-cache-size=512 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-deprecated-copy -Wno-unused-parameter -Wunused -Wparentheses -Wno-deprecated -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS  -fPIC  -MMD -MF tmp/el8_ppc64le_gcc12/src/CommonTools/TrackerMap/src/CommonToolsTrackerMap/TrackerMap.cc.d /scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/src/TrackerMap.cc -o tmp/el8_ppc64le_gcc12/src/CommonTools/TrackerMap/src/CommonToolsTrackerMap/TrackerMap.cc.o
/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/src/TrackerMap.cc: In member function 'void TrackerMap::init()':
  /scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/src/TrackerMap.cc:498:29: warning: array subscript -2 is below array bounds of 'int [43]' [-Warray-bounds]
   498 |         ntotRing[layer_g - 1] = nrings;
      |         ~~~~~~~~~~~~~~~~~~~~^
In file included from /scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/src/TrackerMap.cc:1:
/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/interface/TrackerMap.h:654:7: note: while referencing 'TrackerMap::ntotRing'
  654 |   int ntotRing[43];
      |       ^~~~~~~~
  /scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/src/TrackerMap.cc:499:30: warning: array subscript -2 is below array bounds of 'int [43]' [-Warray-bounds]
   499 |         firstRing[layer_g - 1] = 1;
      |         ~~~~~~~~~~~~~~~~~~~~~^
/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/interface/TrackerMap.h:655:7: note: while referencing 'TrackerMap::firstRing'
  655 |   int firstRing[43];
      |       ^~~~~~~~~
  /scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/src/TrackerMap.cc:498:29: warning: array subscript -2 is below array bounds of 'int [43]' [-Warray-bounds]
   498 |         ntotRing[layer_g - 1] = nrings;
      |         ~~~~~~~~~~~~~~~~~~~~^
/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/interface/TrackerMap.h:654:7: note: while referencing 'TrackerMap::ntotRing'
  654 |   int ntotRing[43];
      |       ^~~~~~~~
  /scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/src/TrackerMap.cc:499:30: warning: array subscript -2 is below array bounds of 'int [43]' [-Warray-bounds]
   499 |         firstRing[layer_g - 1] = 1;
      |         ~~~~~~~~~~~~~~~~~~~~~^
/scratch/cmsbuild/jenkins_b/workspace/build-any-ib/w/tmp/BUILDROOT/7f1a27a7237b5a5b101f4782de85704e/opt/cmssw/el8_ppc64le_gcc12/cms/cmssw/CMSSW_13_3_X_2023-07-27-2300/src/CommonTools/TrackerMap/interface/TrackerMap.h:655:7: note: while referencing 'TrackerMap::firstRing'
  655 |   int firstRing[43];
      |       ^~~~~~~~~
```

--- a/CommonTools/TrackerMap/interface/TrackerMap.h
+++ b/CommonTools/TrackerMap/interface/TrackerMap.h
@@ -1,5 +1,5 @@
-#ifndef _TrackerMap_h_
-#define _TrackerMap_h_
+#ifndef _CommonTools_TrackerMap_TrackerMap_h_
+#define _CommonTools_TrackerMap_TrackerMap_h_
 #include <utility>
 #include <string>
 #include <iostream>
@@ -8,7 +8,6 @@
 #include <cmath>
 #include <map>
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
-#include "FWCore/Framework/interface/ESHandle.h"
 #include "CondFormats/SiStripObjects/interface/SiStripFedCabling.h"
 #include "TColor.h"
 #include <cassert>
@@ -196,42 +195,34 @@ public:
   bool enableLVProcessing;
   bool enableHVProcessing;
   bool tkMapLog;
-  int ndet;   //number of detectors
-  int npart;  //number of detectors parts
   std::string title;
   std::string jsfilename, infilename;
   std::string jsPath;
   bool psetAvailable;
+
   double phival(double x, double y) {
     double phi;
-    double phi1 = atan(y / x);
-    phi = phi1;
-    if (y < 0. && x > 0)
-      phi = phi1 + 2. * M_PI;
-    if (x < 0.)
-      phi = phi1 + M_PI;
-    if (fabs(y) < 0.000001 && x > 0)
-      phi = 0;
-    if (fabs(y) < 0.000001 && x < 0)
-      phi = M_PI;
-    if (fabs(x) < 0.000001 && y > 0)
-      phi = M_PI / 2.;
-    if (fabs(x) < 0.000001 && y < 0)
-      phi = 3. * M_PI / 2.;
-
+    if (fabs(x) >= 0.000001 && fabs(y) >= 0.000001) {
+      phi = atan(y / x);
+      if (y < 0. && x > 0)
+        phi += 2. * M_PI;
+      else if (x < 0.)
+        phi += M_PI;
+    } else if (fabs(y) < 0.000001)
+      phi = x > 0 ? 0 : M_PI;
+    else  // if (fabs(x) < 0.000001)
+      phi = y > 0 ? M_PI / 2. : 3. * M_PI / 2.;
     return phi;
   }
 
   int find_layer(int ix, int iy) {
-    int add;
     int layer = 0;
     if (iy <= xsize) {  //endcap+z
-      add = 15;
+      const int add = 15;
       layer = ix / ysize;
       layer = layer + add + 1;
-    }
-    if (iy > xsize && iy < 3 * xsize) {  //barrel
-      add = 30;
+    } else if (iy < 3 * xsize) {  //barrel
+      const int add = 30;
       if (ix < 2 * ysize) {
         layer = 1;
       } else {
@@ -242,8 +233,7 @@ public:
           layer = layer * 2;
       }
       layer = layer + add;
-    }
-    if (iy >= 3 * xsize) {  //endcap-z
+    } else if (iy >= 3 * xsize) {  //endcap-z
       layer = ix / ysize;
       layer = 15 - layer;
     }
@@ -275,6 +265,7 @@ public:
     }
     return (ncomponent);
   }
+
   double xdpixel(double x) {
     double res;
     if (saveAsSingleLayer)
@@ -506,47 +497,42 @@ public:
 
   int getringCount(int subdet, int partdet, int layer) {
     int ncomponent = 0;
-    if (subdet == 1) {                     //1=pixel
-      if (partdet == 1 || partdet == 3) {  //end-cap
+    if (subdet == 1) {                   //1=pixel
+      if (partdet == 1 || partdet == 3)  //end-cap
         ncomponent = 7;
-      } else {
+      else  //barrel
         ncomponent = 8;
-      }  //barrel
-    }
-    if (subdet == 2) {                     //inner-silicon
-      if (partdet == 1 || partdet == 3) {  //end-cap
+    } else if (subdet == 2) {            //inner-silicon
+      if (partdet == 1 || partdet == 3)  //end-cap
         ncomponent = 3;
-      } else {
+      else  //barrel
         ncomponent = 12;
-      }  //barrel
-    }
-    if (subdet == 3) {     //outer-silicon
-      if (partdet == 1) {  //end-cap-z
+    } else if (subdet == 3) {  //outer-silicon
+      if (partdet == 1) {      //end-cap-z
         if (layer == 1)
           ncomponent = 4;
-        if (layer == 2 || layer == 3)
+        else if (layer == 2 || layer == 3)
           ncomponent = 5;
-        if (layer == 4 || layer == 5 || layer == 6)
+        else if (layer == 4 || layer == 5 || layer == 6)
           ncomponent = 6;
-        if (layer == 7 || layer == 8 || layer == 9)
+        else if (layer == 7 || layer == 8 || layer == 9)
           ncomponent = 7;
-      }
-      if (partdet == 3) {  //endcap+z
+      } else if (partdet == 3) {  //endcap+z
         if (layer == 9)
           ncomponent = 4;
-        if (layer == 8 || layer == 7)
+        else if (layer == 8 || layer == 7)
           ncomponent = 5;
-        if (layer == 6 || layer == 5 || layer == 4)
+        else if (layer == 6 || layer == 5 || layer == 4)
           ncomponent = 6;
-        if (layer == 3 || layer == 2 || layer == 1)
+        else if (layer == 3 || layer == 2 || layer == 1)
           ncomponent = 7;
-      }
-      if (partdet == 2) {  //barrel
+      } else if (partdet == 2) {  //barrel
         ncomponent = 12;
       }
     }
     return (ncomponent);
   }
+
   int getmoduleCount(int subdet, int partdet, int layer, int ring) {
     int ncomponent = 0;
     int spicchif[] = {24, 24, 40, 56, 40, 56, 80};
@@ -571,6 +557,7 @@ public:
     }
     return (ncomponent);
   }
+
   static int layerno(int subdet, int leftright, int layer) {
     if (subdet == 6 && leftright == 1)
       return (10 - layer);
@@ -604,6 +591,7 @@ public:
         return true;
     return false;
   }
+
   int nlayer(int det, int part, int lay) {
     if (det == 3 && part == 1)
       return lay;
@@ -623,7 +611,7 @@ public:
       return lay + 33;
     if (det == 3 && part == 2)
       return lay + 37;
-    return -1;
+    return -1;  // This can never happen: det and part are comprised between 1 and 3
   }
 
   std::string layername(int layer) {
@@ -651,6 +639,7 @@ public:
     s = ons.str();
     return s;
   }
+
   int ntotRing[43];
   int firstRing[43];
 

--- a/CommonTools/TrackerMap/src/TrackerMap.cc
+++ b/CommonTools/TrackerMap/src/TrackerMap.cc
@@ -484,8 +484,8 @@ void TrackerMap::init() {
   gminvalue = 0.;
   gmaxvalue = 0.;  //default global range for online rendering
 
-  ndet = 3;   // number of detectors: pixel, inner silicon, outer silicon
-  npart = 3;  // number of detector parts: endcap -z, barrel, endcap +z
+  int constexpr ndet = 3;   // number of detectors: pixel, inner silicon, outer silicon
+  int constexpr npart = 3;  // number of detector parts: endcap -z, barrel, endcap +z
 
   //allocate module map
   for (int subdet = 1; subdet < ndet + 1; subdet++) {        //loop on subdetectors

